Support "prompt" config setting

vilmibm opened the pull request:
This PR adds support for a new configuration setting, `prompt`, with supported values `auto` and `never`.

I tried to track down all the commands that prompt and updated:

- pr create
- pr merge
- pr review
- auth login
- auth logout
- auth refresh
- issue create
- repo create
- repo fork


I didn't add new tests for the most part since I was just tweaking the code paths already in place for non-tty support.

It's also worth noting that this cannot be configured per-host, something I call out in the default config comments.

Closes #1587

vilmibm committed:
support Config in IOStreams

vilmibm committed:
Add prompt config setting and support in CanPrompt

tranthanhgiang145 commented:
From 93e8736682dee803b618fb2a6e7cfc61d0615160 Mon Sep 17 00:00:00 2001
From: vilmibm <vilmibm@neongrid.space>
Date: Tue, 8 Sep 2020 11:31:35 -0500
Subject: [PATCH 1/7] add prompts to default config

---
 internal/config/config_file_test.go |  2 +-
 internal/config/config_type.go      | 11 +++++++++++
 internal/config/config_type_test.go |  8 ++++++--
 pkg/cmd/config/config.go            |  1 +
 4 files changed, 19 insertions(+), 3 deletions(-)

diff --git a/internal/config/config_file_test.go b/internal/config/config_file_test.go
index 2d7fd4b73..aee767460 100644
--- a/internal/config/config_file_test.go
+++ b/internal/config/config_file_test.go
@@ -110,7 +110,7 @@ github.com:
 	_, err := ParseConfig("config.yml")
 	assert.Nil(t, err)
 
-	expectedMain := "# What protocol to use when performing git operations. Supported values: ssh, https\ngit_protocol: https\n# What editor gh should run when creating issues, pull requests, etc. If blank, will refer to environment.\neditor:\n# Aliases allow you to create nicknames for gh commands\naliases:\n    co: pr checkout\n"
+	expectedMain := "# What protocol to use when performing git operations. Supported values: ssh, https\ngit_protocol: https\n# What editor gh should run when creating issues, pull requests, etc. If blank, will refer to environment.\neditor:\n# Whether gh should use interactive prompts to gather user input. These interactive prompts can cause problems with screen readers and can be disabled by setting this to 'never'\nprompts: auto\n# Aliases allow you to create nicknames for gh commands\naliases:\n    co: pr checkout\n"
 	expectedHosts := `github.com:
     user: keiyuri
     oauth_token: "123456"
diff --git a/internal/config/config_type.go b/internal/config/config_type.go
index 8bb20df43..9c8051666 100644
--- a/internal/config/config_type.go
+++ b/internal/config/config_type.go
@@ -167,6 +167,15 @@ func NewBlankRoot() *yaml.Node {
 						Kind:  yaml.ScalarNode,
 						Value: "",
 					},
+					{
+						HeadComment: "Whether gh should use interactive prompts to gather user input. These interactive prompts can cause problems with screen readers and can be disabled by setting this to 'never'",
+						Kind:        yaml.ScalarNode,
+						Value:       "prompts",
+					},
+					{
+						Kind:  yaml.ScalarNode,
+						Value: "auto",
+					},
 					{
 						HeadComment: "Aliases allow you to create nicknames for gh commands",
 						Kind:        yaml.ScalarNode,
@@ -476,6 +485,8 @@ func defaultFor(key string) string {
 	switch key {
 	case "git_protocol":
 		return defaultGitProtocol
+	case "prompts":
+		return "auto"
 	default:
 		return ""
 	}
diff --git a/internal/config/config_type_test.go b/internal/config/config_type_test.go
index df33effa3..31f6d6619 100644
--- a/internal/config/config_type_test.go
+++ b/internal/config/config_type_test.go
@@ -19,7 +19,7 @@ func Test_fileConfig_Set(t *testing.T) {
 	assert.NoError(t, c.Set("github.com", "user", "hubot"))
 	assert.NoError(t, c.Write())
 
-	expected := "# What protocol to use when performing git operations. Supported values: ssh, https\ngit_protocol: https\n# What editor gh should run when creating issues, pull requests, etc. If blank, will refer to environment.\neditor: nano\n# Aliases allow you to create nicknames for gh commands\naliases:\n    co: pr checkout\n"
+	expected := "# What protocol to use when performing git operations. Supported values: ssh, https\ngit_protocol: https\n# What editor gh should run when creating issues, pull requests, etc. If blank, will refer to environment.\neditor: nano\n# Whether gh should use interactive prompts to gather user input. These interactive prompts can cause problems with screen readers and can be disabled by setting this to 'never'\nprompts: auto\n# Aliases allow you to create nicknames for gh commands\naliases:\n    co: pr checkout\n"
 	assert.Equal(t, expected, mainBuf.String())
 	assert.Equal(t, `github.com:
     git_protocol: ssh
@@ -37,7 +37,7 @@ func Test_defaultConfig(t *testing.T) {
 	cfg := NewBlankConfig()
 	assert.NoError(t, cfg.Write())
 
-	expected := "# What protocol to use when performing git operations. Supported values: ssh, https\ngit_protocol: https\n# What editor gh should run when creating issues, pull requests, etc. If blank, will refer to environment.\neditor:\n# Aliases allow you to create nicknames for gh commands\naliases:\n    co: pr checkout\n"
+	expected := "# What protocol to use when performing git operations. Supported values: ssh, https\ngit_protocol: https\n# What editor gh should run when creating issues, pull requests, etc. If blank, will refer to environment.\neditor:\n# Whether gh should use interactive prompts to gather user input. These interactive prompts can cause problems with screen readers and can be disabled by setting this to 'never'\nprompts: auto\n# Aliases allow you to create nicknames for gh commands\naliases:\n    co: pr checkout\n"
 	assert.Equal(t, expected, mainBuf.String())
 	assert.Equal(t, "", hostsBuf.String())
 
@@ -49,6 +49,10 @@ func Test_defaultConfig(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, "", editor)
 
+	prompt, err := cfg.Get("", "prompts")
+	assert.Nil(t, err)
+	assert.Equal(t, "auto", prompt)
+
 	aliases, err := cfg.Aliases()
 	assert.Nil(t, err)
 	assert.Equal(t, len(aliases.All()), 1)
diff --git a/pkg/cmd/config/config.go b/pkg/cmd/config/config.go
index 3fa86ddcb..0d4c6cdb9 100644
--- a/pkg/cmd/config/config.go
+++ b/pkg/cmd/config/config.go
@@ -18,6 +18,7 @@ func NewCmdConfig(f *cmdutil.Factory) *cobra.Command {
 			Current respected settings:
 			- git_protocol: "https" or "ssh". Default is "https".
 			- editor: if unset, defaults to environment variables.
+			- prompts: "auto" or "never". Default is "auto".
 		`),
 	}
 

From 6f4dd5962c0ca40903af12d7a2927e302e800677 Mon Sep 17 00:00:00 2001
From: vilmibm <vilmibm@neongrid.space>
Date: Tue, 8 Sep 2020 11:46:54 -0500
Subject: [PATCH 2/7] respect prompt cfg in pr create

---
 pkg/cmd/pr/create/create.go | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/pkg/cmd/pr/create/create.go b/pkg/cmd/pr/create/create.go
index 6b144a5e9..618446a8c 100644
--- a/pkg/cmd/pr/create/create.go
+++ b/pkg/cmd/pr/create/create.go
@@ -241,7 +241,17 @@ func createRun(opts *CreateOptions) error {
 		Milestones: milestoneTitles,
 	}
 
-	interactive := isTerminal && !(opts.TitleProvided && opts.BodyProvided)
+	cfg, err := opts.Config()
+	if err != nil {
+		return err
+	}
+
+	prompt, err := cfg.Get(headRepo.RepoHost(), "prompts")
+	if err != nil {
+		return err
+	}
+
+	interactive := isTerminal && !(opts.TitleProvided && opts.BodyProvided) && (prompt != "never")
 
 	if !opts.WebMode && !opts.Autofill && interactive {
 		var nonLegacyTemplateFiles []string
@@ -312,10 +322,6 @@ func createRun(opts *CreateOptions) error {
 	// In either case, we want to add the head repo as a new git remote so we
 	// can push to it.
 	if headRemote == nil {
-		cfg, err := opts.Config()
-		if err != nil {
-			return err
-		}
 		cloneProtocol, _ := cfg.Get(headRepo.RepoHost(), "git_protocol")
 
 		headRepoURL := ghrepo.FormatRemoteURL(headRepo, cloneProtocol)

From f050e34e7715a458afd59425f9ec990779a8b790 Mon Sep 17 00:00:00 2001
From: vilmibm <vilmibm@neongrid.space>
Date: Tue, 8 Sep 2020 14:59:19 -0500
Subject: [PATCH 3/7] avoid raw string

---
 internal/config/config_type.go | 5 ++++-
 pkg/cmd/pr/create/create.go    | 2 +-
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/internal/config/config_type.go b/internal/config/config_type.go
index 9c8051666..bb20db398 100644
--- a/internal/config/config_type.go
+++ b/internal/config/config_type.go
@@ -10,7 +10,10 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
-const defaultGitProtocol = "https"
+const (
+	defaultGitProtocol = "https"
+	NeverPrompt        = "never"
+)
 
 // This interface describes interacting with some persistent configuration for gh.
 type Config interface {
diff --git a/pkg/cmd/pr/create/create.go b/pkg/cmd/pr/create/create.go
index 618446a8c..7e8b81214 100644
--- a/pkg/cmd/pr/create/create.go
+++ b/pkg/cmd/pr/create/create.go
@@ -251,7 +251,7 @@ func createRun(opts *CreateOptions) error {
 		return err
 	}
 
-	interactive := isTerminal && !(opts.TitleProvided && opts.BodyProvided) && (prompt != "never")
+	interactive := isTerminal && !(opts.TitleProvided && opts.BodyProvided) && (prompt != config.NeverPrompt)
 
 	if !opts.WebMode && !opts.Autofill && interactive {
 		var nonLegacyTemplateFiles []string

From c8d590446beb56b8afbf43bcbdd972e5d0b5bd71 Mon Sep 17 00:00:00 2001
From: vilmibm <vilmibm@neongrid.space>
Date: Tue, 8 Sep 2020 15:47:59 -0500
Subject: [PATCH 4/7] use baserepo, add constant

---
 internal/config/config_type.go | 3 ++-
 pkg/cmd/pr/create/create.go    | 2 +-
 2 files changed, 3 insertions(+), 2 deletions(-)

diff --git a/internal/config/config_type.go b/internal/config/config_type.go
index bb20db398..a91567a04 100644
--- a/internal/config/config_type.go
+++ b/internal/config/config_type.go
@@ -13,6 +13,7 @@ import (
 const (
 	defaultGitProtocol = "https"
 	NeverPrompt        = "never"
+	AutoPrompt         = "auto"
 )
 
 // This interface describes interacting with some persistent configuration for gh.
@@ -489,7 +490,7 @@ func defaultFor(key string) string {
 	case "git_protocol":
 		return defaultGitProtocol
 	case "prompts":
-		return "auto"
+		return AutoPrompt
 	default:
 		return ""
 	}
diff --git a/pkg/cmd/pr/create/create.go b/pkg/cmd/pr/create/create.go
index 7e8b81214..72c348b1b 100644
--- a/pkg/cmd/pr/create/create.go
+++ b/pkg/cmd/pr/create/create.go
@@ -246,7 +246,7 @@ func createRun(opts *CreateOptions) error {
 		return err
 	}
 
-	prompt, err := cfg.Get(headRepo.RepoHost(), "prompts")
+	prompt, err := cfg.Get(baseRepo.RepoHost(), "prompts")
 	if err != nil {
 		return err
 	}

From 18ce9059573ef618dda498fd0e4e86b806206d79 Mon Sep 17 00:00:00 2001
From: Sam Coe <samcoe@github.com>
Date: Wed, 9 Sep 2020 12:13:09 +0200
Subject: [PATCH 5/7] Obey prompts config for repo create

---
 pkg/cmd/repo/create/create.go | 21 +++++++++++++++------
 1 file changed, 15 insertions(+), 6 deletions(-)

diff --git a/pkg/cmd/repo/create/create.go b/pkg/cmd/repo/create/create.go
index fc53b460e..2343df301 100644
--- a/pkg/cmd/repo/create/create.go
+++ b/pkg/cmd/repo/create/create.go
@@ -147,6 +147,21 @@ func createRun(opts *CreateOptions) error {
 		isVisibilityPassed = true
 	}
 
+	cfg, err := opts.Config()
+	if err != nil {
+		return err
+	}
+
+	prompts, err := cfg.Get("", "prompts")
+	if err != nil {
+		return err
+	}
+
+	if (prompts == config.NeverPrompt) && (!isNameAnArg || !isVisibilityPassed) {
+		// TODO: Create better error message(s) when name and/or visibility are not present
+		return fmt.Errorf("provide name and visibility flags")
+	}
+
 	// Trigger interactive prompt if name is not passed
 	if !isNameAnArg {
 		newName, newDesc, newVisibility, err := interactiveRepoCreate(isDescEmpty, isVisibilityPassed, repoToCreate.RepoName())
@@ -245,11 +260,6 @@ func createRun(opts *CreateOptions) error {
 			fmt.Fprintln(stdout, repo.URL)
 		}
 
-		// TODO This is overly wordy and I'd like to streamline this.
-		cfg, err := opts.Config()
-		if err != nil {
-			return err
-		}
 		protocol, err := cfg.Get(repo.RepoHost(), "git_protocol")
 		if err != nil {
 			return err
@@ -346,7 +356,6 @@ func interactiveRepoCreate(isDescEmpty bool, isVisibilityPassed bool, repoName s
 	}
 
 	return answers.RepoOwner, answers.RepoDescription, strings.ToUpper(answers.RepoVisibility), nil
-
 }
 
 func confirmSubmission(repoName string, repoOwner string, isConfirmFlagPassed *bool) (bool, error) {

From 191e0e89f9665757cea17065dc81306ae3e04738 Mon Sep 17 00:00:00 2001
From: Sam Coe <samcoe@github.com>
Date: Wed, 9 Sep 2020 12:28:06 +0200
Subject: [PATCH 6/7] Obey prompts config for issue create

---
 pkg/cmd/issue/create/create.go | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/pkg/cmd/issue/create/create.go b/pkg/cmd/issue/create/create.go
index 8d47e4e27..402809f6e 100644
--- a/pkg/cmd/issue/create/create.go
+++ b/pkg/cmd/issue/create/create.go
@@ -148,7 +148,17 @@ func createRun(opts *CreateOptions) error {
 	title := opts.Title
 	body := opts.Body
 
-	interactive := !(opts.TitleProvided && opts.BodyProvided)
+	cfg, err := opts.Config()
+	if err != nil {
+		return err
+	}
+
+	prompt, err := cfg.Get("", "prompts")
+	if err != nil {
+		return err
+	}
+
+	interactive := !(opts.TitleProvided && opts.BodyProvided) && (prompt != config.NeverPrompt)
 
 	if interactive && !isTerminal {
 		return fmt.Errorf("must provide --title and --body when not attached to a terminal")

From a132a0f809032bf5e91944cc1b22c1b611062da7 Mon Sep 17 00:00:00 2001
From: Sam Coe <samcoe@github.com>
Date: Wed, 9 Sep 2020 13:00:21 +0200
Subject: [PATCH 7/7] Obey prompts config for pr review

---
 pkg/cmd/pr/review/review.go | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/pkg/cmd/pr/review/review.go b/pkg/cmd/pr/review/review.go
index 19f33ca84..8ba65e22a 100644
--- a/pkg/cmd/pr/review/review.go
+++ b/pkg/cmd/pr/review/review.go
@@ -103,6 +103,21 @@ func NewCmdReview(f *cmdutil.Factory, runF func(*ReviewOptions) error) *cobra.Co
 				}
 			}
 
+			cfg, err := opts.Config()
+			if err != nil {
+				return err
+			}
+
+			prompt, err := cfg.Get("", "prompts")
+			if err != nil {
+				return err
+			}
+
+			if (prompt == config.NeverPrompt) && found != 1 {
+				// TODO: Create better error message(s)
+				return &cmdutil.FlagError{Err: errors.New("need exactly one of --approve, --request-changes, or --comment")}
+			}
+
 			if found == 0 && opts.Body == "" {
 				if !opts.IO.IsStdoutTTY() || !opts.IO.IsStdinTTY() {
 					return &cmdutil.FlagError{Err: errors.New("--approve, --request-changes, or --comment required when not attached to a tty")}

vilmibm committed:
split auth flow code into new internal package

vilmibm committed:
update default config test values

vilmibm committed:
use CanPrompt in commands

mislav commented on pkg/iostreams/iostreams.go:
```diff
@@ -135,7 +139,7 @@ func (s *IOStreams) ColorScheme() *ColorScheme {
 }
 
-func System() *IOStreams {
+func System(configFunc func() (config.Config, error)) *IOStreams {
```
So far, IOStreams was pretty application-agnostic. Could we possibly keep that approach and avoid coupling it to Config?

IOStreams can still absolutely react to user's configuration values, but maybe Config and IOStreams should have a facilitator between them that takes preferences from one and applies them to the other.

Currently, that facilitator can be `main.go`, as it both instantiates `iostreams.System()` (through `factory.New()`) and has access to the user's config at the same time (for expanding aliases and doing an auth check). When `main.go` detects that a user does not wish to see prompts, it could simply invoke something like `IO.SetCanPrompt(false)`. Would that make sense?

vilmibm replied:
yep, i agree that would be better

vilmibm committed:
do not pass entire config to iostreams

Commit:
prompts disabled test for auth refresh

the behavior for nontty vs. prompts disabled is pretty different so i
opted for a new test, here

ampinsk commented:
One nit: the values of `auto` and `never` weren't very intuitive to me. How about `enabled` and `disabled`?

mislav commented on pkg/cmd/auth/login/login.go:
```diff
@@ -71,6 +72,12 @@ func NewCmdLogin(f *cmdutil.Factory, runF func(*LoginOptions) error) *cobra.Comm
+				if !opts.IO.CanPrompt() {
+					// TODO this error message kind of sucks since the prompting might be disabled from being
+					// in a nontty or from the config setting. Is it enough info?
+					return &cmdutil.FlagError{Err: errors.New("--with-token required when prompts disabled")}
```
I think we should avoid saying “when prompts disabled” because `CanPrompt` can return false even if the user hasn't explicitly disabled prompts. Maybe `--with-token required when not running in interactive mode`?

mislav commented on pkg/cmd/auth/logout/logout.go:
```diff
@@ -152,6 +148,8 @@ func logoutRun(opts *LogoutOptions) error {
 		return fmt.Errorf("failed to write config, authentication configuration not updated: %w", err)
 	}
 
+	isTTY := opts.IO.IsStdinTTY() && opts.IO.IsStdoutTTY()
```
I wonder if we could also name these tty checks something more descriptive that declares intent, like `opts.IO.ShouldShowConfirmation()` (ideally with a better name). Perhaps in a separate changeset!

vilmibm replied:
I like the idea of an `IO.ShowConfirmations()`! i'll open an issue for it.

mislav commented on pkg/cmd/auth/refresh/refresh.go:
```diff
@@ -48,6 +50,17 @@ func NewCmdRefresh(f *cmdutil.Factory, runF func(*RefreshOptions) error) *cobra.
+
+			if opts.Hostname == "" && !opts.IO.CanPrompt() {
+				// here, we know we are attached to a TTY but prompts are disabled
+				return &cmdutil.FlagError{Err: errors.New("--hostname  required when prompts disabled")}
```
Nit: double space

mislav commented on pkg/cmd/issue/create/create.go:
```diff
@@ -157,6 +157,10 @@ func createRun(opts *CreateOptions) error {
 	}
 
+	if interactive && !opts.IO.CanPrompt() {
+		return fmt.Errorf("must provide --title and --body when prompts are disabled")
```
Nit: if both of these checks are only dependent on the state of `--title` and `--body` as provided by flags, could we move them to the flag-parsing section, e.g. the RunE block? This would ensure that we error out as soon as we have the information that the command cannot proceed

mislav commented on pkg/cmd/repo/create/create.go:
```diff
@@ -73,6 +73,15 @@ func NewCmdCreate(f *cmdutil.Factory, runF func(*CreateOptions) error) *cobra.Co
+					return &cmdutil.FlagError{Err: errors.New("must pass name argument when prompts disabled")}
+				}
+				if !opts.Internal && !opts.Private && !opts.Public {
+					opts.Public = true
```
I strongly feel that we should either default to `--private` or error out in non-interactive mode when repo visibility wasn't explicitly declared.

vilmibm replied:
defaulting to private feels right to me, but didn't we get multiple issues complaining about not defaulting to public for repo creation? or am I misremembering?

vilmibm replied:
I set it to error now, happy to update to something else later. I figured the annoyance of getting an error is better than the annoyance of having to go to the browser and toggle the repo visibility if the default isn't what you wanted.

mislav commented on cmd/gh/main.go:
```diff
@@ -45,6 +45,18 @@ func main() {
+	cfg, err := cmdFactory.Config()
+	if err != nil {
+		fmt.Fprintf(stderr, "failed to read configuration:  %s\n", err)
+		os.Exit(2)
```
Are there legitimate cases when configuration could not be read, but we want to allow executing commands? E.g. when the config file has a syntax error due to manual edits?

I can't think of any, but if you do think of some cases, it's always an option to write a warning notice to stderr here instead of failing. Right now, I'm fine with failing, since a broken config file means that authentication info could not be read, and a missing config file does not raise any errors.

Nit: since you have it here, you can now avoid reading configuration in the lower `authCheckEnabled` block.

vilmibm replied:
I can't think of any either and would prefer to fail; running with a broken config can put users into a place that is very hard for them and us to debug

mislav commented on cmd/gh/main.go:
```diff
@@ -45,6 +45,18 @@ func main() {
+
+	prompt, _ := cfg.Get("", "prompt")
+
+	if prompt == config.NeverPrompt {
+		cmdFactory.IOStreams.SetNeverPrompt(true)
```
Love how easy & simple this looks in the end; thanks for the hard work!! 💖

mislav commented on internal/authflow/flow.go:
```diff
@@ -22,14 +23,9 @@ var (
 )
 
-// IsGitHubApp reports whether an OAuth app is "GitHub CLI" or "GitHub CLI (dev)"
-func IsGitHubApp(id string) bool {
```
🔥 🔥 nice find!

mislav reviewed (approved):
Nice work! This looks good; I've pointed out some minor things that I see as regressions in non-interactive mode.

The only overall direction I would change is avoiding saying “when prompts are disabled” in error messages when `CanPrompt()` returns false, because those messages will also be shown in CI/CD environments and scripts where prompts might have not been disabled per se, but are not available due to lack of terminal. So maybe a better message would be to say “when running non-interactively” or something similar?

vilmibm committed:
review feedback

vilmibm committed:
reuse cfg reference

vilmibm committed:
Merge branch 'disable-prompts' of github.com:cli/cli into disable-prompts

vilmibm committed:
do not assume public when creating repo non-interactively
